### PR TITLE
RStudio Helm chart

### PR DESCRIPTION
**Pull request by yonghaoy**
```
 helm install  --set serviceAccount.annotations.gcpServiceAccount="yonghaotest@cryptic-skyline-320001.iam.gserviceaccount.com" aou-rstudio-chart-setup ./terra-app-setup-chart
```

```
helm install aou-rstudio-chart aou-rstudio-chart  --values aou-rstudio-chart/test_values.yaml 
```

Wondershaper: 2MB/s = 16Mitb/s

Screenshot 2023-05-01 at 7 01 09 PM

### CODE REVIEW

File: terra-app-helm/aou-rstudio-chart/templates/deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+          livenessProbe:
+            httpGet:
+              path: \
+              port: http
+          readinessProbe:
+            httpGet:
+              path: \
+              port: http
+          volumeMounts:
+            - mountPath: "{{ .Values.persistence.mountPath -}}"
```
Comment (rtitle) [inline terra-app-helm/aou-rstudio-chart/templates/deployment.yaml:123]: I _think_ welder wants the `mountPath` to be `/work` not `/home/rstudio`: https://github.com/DataBiosphere/welder/blob/6648f5cc6ff170b7413c2561e3a8eb1bce76ab6c/server/src/main/resources/reference.conf#LL6C1-L6C1

File: terra-app-helm/aou-rstudio-chart/templates/deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,138 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: {{ include "app.fullname" . }}
+  labels:
+    {{- include "app.labels" . | nindent 4 }}
+spec:
+  {{- if not .Values.autoscaling.enabled }}
+  replicas: {{ .Values.replicaCount }}
```
Comment (rtitle) [inline terra-app-helm/aou-rstudio-chart/templates/deployment.yaml:9]: I guess some of this is copy-pasted.. but as this is a single-user RStudio deployment I'm guessing we'd always want `replicaCount=1` and not enable HPA?

File: terra-app-helm/aou-rstudio-chart/templates/deployment.yaml [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+          imagePullPolicy: "IfNotPresent"
+          env:
+            - name: CLOUD_PROVIDER
+              value: "gcp"
+            - name: LOCKING_ENABLED
+              value: "true"
+            - name: SHOULD_BACKGROUND_SYNC
+              value: "true"
+            - name: WSM_URL
+              value: "dummy"
```
Comment (rtitle) [inline terra-app-helm/aou-rstudio-chart/templates/deployment.yaml:83]: Probably want to make these helm values so Leo can inject them I think..

File: terra-app-helm/aou-rstudio-chart/templates/deployment.yaml [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+            - {{ . | quote }}
+            {{- end }}
+          env:
+            {{- range .Values.extraEnv }}
+            - name: {{ .name }}
+              value: {{ .value | quote }}
+            {{- end }}
+          ports:
+            - name: http
+              containerPort: {{ .Values.image.port }}
```
Comment (rtitle) [inline terra-app-helm/aou-rstudio-chart/templates/deployment.yaml:57]: Looks like this is 8787 in values.yaml but I see 8080 elsewhere in this file (L93, 112)

File: terra-app-helm/aou-rstudio-chart/templates/hpa.yaml [hunk 5]
```diff
@@ -0,0 +1,28 @@
+{{- if .Values.autoscaling.enabled }}
```
Comment (rtitle) [inline terra-app-helm/aou-rstudio-chart/templates/hpa.yaml:1]: Would we ever enable HPA in a single-user RStudio deployment?

File: terra-app-helm/aou-rstudio-chart/templates/persistence.yaml [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+metadata:
+  name: {{ include "app.fullname" . }}-pv
+spec:
+  storageClassName: "{{ .Values.persistence.storageClass }}"
+  capacity:
+    storage: {{ .Values.persistence.size }}
+  accessModes:
+    - {{ .Values.persistence.accessMode }}
+  {{- if .Values.persistence.hostPath }}
+  hostPath:
```
Comment (rtitle) [inline terra-app-helm/aou-rstudio-chart/templates/persistence.yaml:13]: Probably don't want to use `hostPath`, it's really just for testing on a single node cluster: https://kubernetes.io/docs/tasks/configure-pod-container/configure-persistent-volume-storage/#create-a-persistentvolume

File: terra-app-helm/aou-rstudio-chart/templates/persistence.yaml [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+---
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: {{ include "app.fullname" . }}-pvc
+spec:
+  storageClassName: "{{ .Values.persistence.storageClass }}"
+  volumeName: {{ include "app.fullname" . }}-pv
+  accessModes:
+    - {{ .Values.persistence.accessMode }}
```
Comment (rtitle) [inline terra-app-helm/aou-rstudio-chart/templates/persistence.yaml:33]: Just noting a lot of this is highly templated which makes it a bit confusing.. consider removing templates and just putting in raw values where it makes sense. E.g. RStudio probably always wants `accessMode: ReadWriteMany`, `storageClass: default`, etc

File: terra-app-helm/aou-rstudio-chart/templates/hpa.yaml (hunk 5 above)
Comment (yonghaoy) [reply terra-app-helm/aou-rstudio-chart/templates/hpa.yaml:1]: good point. Those were auto created by `helm create`

File: terra-app-setup-chart/templates/service-account.yaml [hunk 8]
```diff
@@ -1,7 +1,7 @@
 apiVersion: v1
 kind: ServiceAccount
 metadata:
-  name: {{ .Values.serviceAccount.name | quote }}
```
Comment (yonghaoy) [inline terra-app-setup-chart/templates/service-account.yaml:4]: will revert this back after testing

File: terra-app-helm/aou-rstudio-chart/templates/deployment.yaml [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+        fsGroupChangePolicy: "OnRootMismatch"
+      volumes:
+        - name: {{ include "app.fullname" . }}-storage
+          persistentVolumeClaim:
+            claimName: {{ include "app.fullname" . }}-pvc
+      containers:
+        - name: aou-rstudio
+          image: "{{ .Values.image.image }}"
+          imagePullPolicy: IfNotPresent
+          securityContext:
```
Comment (LizBaldo) [inline terra-app-helm/aou-rstudio-chart/templates/deployment.yaml:37]: Thanks! We will close this ticket then :) https://broadworkbench.atlassian.net/browse/IA-4115

File: terra-app-helm/aou-rstudio-chart/templates/deployment.yaml [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+            httpGet:
+              path: "{{ .Values.image.baseUrl }}"
+              port: http
+          readinessProbe:
+            httpGet:
+              path: "{{ .Values.image.baseUrl }}"
+              port: http
+          resources: {}
+        - name: welder
+          image: "us.gcr.io/broad-dsp-gcr-public/welder-server:6648f5c"
```
Comment (LizBaldo) [inline terra-app-helm/aou-rstudio-chart/templates/deployment.yaml:65]: Thinking out loud here, do you want the chart to keep up with the latest working version of welder or is pointing to 6648f5c enough? Looks like this has been stable for the past 6 months so not too worried

File: terra-app-helm/aou-rstudio-chart/templates/deployment.yaml [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+              value: "true"
+            - name: SHOULD_BACKGROUND_SYNC
+              value: "true"
+            - name: STORAGE_CONTAINER_RESOURCE_ID
+              value: "dummy"
+            - name: STAGING_STORAGE_CONTAINER_RESOURCE_ID
+              value: "dummy"
+            - name: PORT
+              value: "8080"
+#            - name: WSM_URL
```
Comment (LizBaldo) [inline terra-app-helm/aou-rstudio-chart/templates/deployment.yaml:80]: We will iterate on that list as part of https://broadworkbench.atlassian.net/browse/IA-4268

You might need extra things like the `SAM_URL`, but could do without `RUNTME_NAME` and maybe `STAGING_BUCKET`

File: terra-app-helm/aou-rstudio-chart/test_values.yaml [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+  enabled: true
+  annotations:
+#   nginx.ingress.kubernetes.io/auth-tls-pass-certificate-to-upstream: "true"
+#   nginx.ingress.kubernetes.io/auth-tls-secret: "{{ $.Release.Namespace }}/ca-secret"
+#   nginx.ingress.kubernetes.io/auth-tls-verify-client: "on"
+#   nginx.ingress.kubernetes.io/auth-tls-verify-depth: "1"
+#   nginx.ingress.kubernetes.io/ssl-redirect: "true"
+#   nginx.ingress.kubernetes.io/proxy-redirect-from: "https://k8s-host/"
+#   nginx.ingress.kubernetes.io/proxy-redirect-to: "https://leo-host/"
+   nginx.ingress.kubernetes.io/rewrite-target: "/$2"
```
[the remark this answers is not in the document]
Comment (LizBaldo) [reply terra-app-helm/aou-rstudio-chart/test_values.yaml:38]: Yep, we will iterate on the final list during the leo implementation

File: terra-app-helm/aou-rstudio-chart/test_values.yaml [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+
+welder:
+  extraEnv:
+  - name: GOOGLE_PROJECT
+    value: "cryptic-skyline-320001"
+  - name: STAGING_BUCKET
+    value: "aou-startup-script"
+  - name: CLUSTER_NAME
+    value: "all-of-us-123"
+  - name: RUNTIME_NAME
```
Comment (LizBaldo) [inline terra-app-helm/aou-rstudio-chart/test_values.yaml:59]: Oh I am curious to see if that was actually needed? We won't be spinning up a runtime from leo

File: terra-app-setup-chart/templates/service-account.yaml (hunk 8 above)
Comment (LizBaldo) [reply terra-app-setup-chart/templates/service-account.yaml:4]: Fells like this can be reverted now 👍

**Comment by LizBaldo** [comment]
@yonghaoy I am going to start working on the leo side of the implementation, keep me posted on where this chart will be located 👍

**Comment by yonghaoy** [comment]
> @yonghaoy I am going to start working on the leo side of the implementation, keep me posted on where this chart will be located 👍

Thanks! I was thinking to keep that open until we figured out the Leo changes. 
Can you launch helm chart using this URL? https://raw.githubusercontent.com/DataBiosphere/terra-app/58f6e41f9842450252f08b5112e19cc241242f27/terra-app-helm/aou-rstudio-chart
If not, I can upload it into a GCS bucket

**Comment by LizBaldo** [comment]
Sounds good, let me try the URL for now :)

**Review by jdcanas — APPROVED** [review]

**Comment by LizBaldo** [comment]
@yonghaoy would it be possible to upload the chart to a bucket so I can pull it from there? I am getting a 404 using the raw github link, and I am unsure if the regular github URL would work

**Comment by LizBaldo** [comment]
That might be a good idea yes, `helm repo add terra-app https://terra-app-charts.storage.googleapis.com` was how I was envisioning it, but that likely won't work until we merge this PR. Let me ask around for advice and get back to you

**Comment by LizBaldo** [comment]
@yonghaoy I think I am unblocked, I will just used the command I linked above to push the chart to `terra-app` :)

**Comment by LizBaldo** [comment]
I ended up creating a new storage bucket and loaded the chart here fyi https://console.cloud.google.com/storage/browser/_details/terra-app-helm/aou-rstudio-chart/aou-rstudio-chart-0.1.0.tgz;tab=live_object?project=broad-dsp-gcr-public

**Comment by LizBaldo** [comment]
@yonghaoy I am getting really close to installing this chart via leo, but I am running into issues pulling the rstudio image due to not having enough permissions (pulling the welder image works though)
`gcr.io/cryptic-skyline-320001/terra-rstudio-aou:smoke-test-1`

**Comment by LizBaldo** [comment]
@Qi77Qi FYI, this is the main PR setting up the RSTUDIO/WELDER chart :)

**Pull request merged**